### PR TITLE
[SNOW-1760495] Reduce test errors on AST enabled merge gate / pass test_dataframe.py

### PR DESCRIPTION
**Pull request by sfc-gh-lspiegelberg**
Fixes several tests to reduce test failures for the merge-gate with `session.ast_enabled=True`. With this PR all tests except for the interval related ones pass for `test_dataframe.py`.

Details:
- Adds `ast_id` to shallow copy protocol for Dataframe.
- Modify APIs so checks happen before AST encoding to pass negative tests.
- Add missing `ast_stmt` to `select(table_function(...))`
- Fix filter test.
- Fix overflow issue when using `datetime.time` due to timezone encoding.
- Support array type (array.py from stdlib) by mapping to list.
- Fix na functions to carry out checks for `subset` parameter.
- Explicitly type check for `col` and `column` as expected in existing tests.
- Add `NotImplementedError` for interval type in `make_interval`.
- Fix wrong usage of Expression in `function.col(...)`, use `Column` instead in `apply_in_pandas`.

### CODE REVIEW

File: src/snowflake/snowpark/dataframe.py [hunk 1]
```diff
@@ -878,7 +879,9 @@ def __copy__(self) -> "DataFrame":
             new_plan._query_params = self._select_statement.query_params
         else:
             new_plan = copy.copy(self._plan)
-        return DataFrame(self._session, new_plan)
+        df = DataFrame(self._session, new_plan)
+        df._ast_id = self._ast_id
```
Comment (sfc-gh-oplaton) [inline src/snowflake/snowpark/dataframe.py:883]: If `DataFrame` never, ever changes in place, this approach is safe. If any meaningful `DataFrame` API mutates the object in place, using the same variable ID for two objects will lead to very surprising results at some point in the future.

We might have this pattern in other places, too.

File: src/snowflake/snowpark/dataframe_na_functions.py [hunk 2]
```diff
@@ -185,10 +197,8 @@ def drop(
         # if subset is not provided, drop will be applied to all columns
         if subset is None:
             subset = self._dataframe.columns
-        elif isinstance(subset, str):
+        if isinstance(subset, str):
```
Comment (sfc-gh-oplaton) [inline src/snowflake/snowpark/dataframe_na_functions.py:200]: This change seems unnecessary, `elif` was doing the correct thing.

File: src/snowflake/snowpark/dataframe_na_functions.py [hunk 3]
```diff
@@ -376,10 +388,8 @@ def fill(
 
         if subset is None:
             subset = self._dataframe.columns
-        elif isinstance(subset, str):
+        if isinstance(subset, str):
```
Comment (sfc-gh-oplaton) [inline src/snowflake/snowpark/dataframe_na_functions.py:391]: Unnecessary?

**Review by sfc-gh-oplaton — APPROVED** [review]
Generally looks good. Minor inconsistencies in handling `subset` when it's `None` vs `str`. 2 functions now use `if` after `if`, the third one uses the previous pattern.

File: src/snowflake/snowpark/dataframe.py [hunk 4]
```diff
@@ -5287,7 +5306,10 @@ def random_split(
             if seed:
                 ast.seed.value = seed
             if statement_params:
-                ast.statement_params = statement_params
+                for k, v in statement_params.items():
```
Comment (sfc-gh-evandenberg) [inline src/snowflake/snowpark/dataframe.py:5309]: Reuse ast_utils.build_expr_from_dict_str_str
```suggestion
                build_expr_from_dict_str_str(ast.statement_params, statement_params)
```

File: src/snowflake/snowpark/dataframe_na_functions.py [hunk 5]
```diff
@@ -63,6 +63,16 @@ def _is_value_type_matching_for_na_function(
...
 
 
+def _check_subset_parameter(subset: Optional[Union[str, Iterable[str]]]) -> None:
+    """Produces exception when invalid subset parameter was passed."""
+    if (
+        subset is not None
+        and not isinstance(subset, str)
+        and not isinstance(subset, (list, tuple))
+    ):
+        raise TypeError("subset should be a list or tuple of column names")
```
Comment (sfc-gh-evandenberg) [inline src/snowflake/snowpark/dataframe_na_functions.py:73]:
```suggestion
        raise TypeError("subset should be a single column name, list or tuple of column names")
```

File: src/snowflake/snowpark/dataframe.py (hunk 1 above)
Comment (sfc-gh-evandenberg) [reply src/snowflake/snowpark/dataframe.py:883]: What is reason for this, performance?

**Review by sfc-gh-evandenberg — APPROVED** [review]
LGTM with some suggestions.

File: src/snowflake/snowpark/dataframe.py (hunk 1 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/dataframe.py:883]: I think this a copy-by-reference vs. deep-copy concern here. Indeed we should be careful regarding this. Atm, this is done using copy-by-reference. However, we may want to add a separate IR API to emulate the copy behavior (maybe reuse cache?)

cc @sfc-gh-azwiegincew 

I'll add a JIRA with todo so we can pickup the copy behavior there.

File: src/snowflake/snowpark/dataframe_na_functions.py (hunk 2 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/dataframe_na_functions.py:200]: good spot, unified to `elif`

File: src/snowflake/snowpark/dataframe_na_functions.py (hunk 3 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/dataframe_na_functions.py:391]: unified to `elif`

File: src/snowflake/snowpark/dataframe_na_functions.py (hunk 5 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/dataframe_na_functions.py]: I like the suggestion, sounds much better. Will have to update existing tests.

**Pull request merged**